File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    // int res = 0;
    int maxPathSum(TreeNode* root) {
        int res = INT_MIN;
        calMax(root,res);
        return res;
    }

    int calMax(TreeNode* node,int& res){
        if(!node) return -10000;
        int lst = calMax(node->left,res);
        int rst = calMax(node->right,res);
        res=max({res,node->val,lst+node->val,rst+node->val,lst+rst+node->val});
        return max(node->val+max(lst,rst),node->val);
    }

};

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:

    def dfs(self,node,res):
        if node is None: 
            return float("-inf") 
        lst = self.dfs(node.left,res)
        rst = self.dfs(node.right,res)
        curr = node.val 
        res[0]=max(res[0],lst,rst,node.val+lst,node.val,node.val+rst,node.val+lst+rst)
        return max(node.val+lst,node.val+rst,node.val)



    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        res = [-10000000]
        self.dfs(root,res)
        return res[-1]